server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -355,6 +356,13 @@ private DfsSearchResult executeDfsPhase(ShardSearchRequest request,
                                             SearchShardTask task,
                                             boolean keepStatesInContext) throws IOException {
         ReaderContext readerContext = createOrGetReaderContext(request, keepStatesInContext);
```

Comment:
Can you add a small comment explaining why it's important to check cancellation here (frozen readers are not opened yet) ?

Reply:
it would be fantastic to have a test for this too, I am guessing it is not easy though. Ideally, removing this code block would make something fail so that further refactoring this code would not end up removing this optimization inadvertently.

Reply:
I tried this: https://github.com/elastic/elasticsearch/pull/69795/commits/717fb97b10b5c701a95f4e87f1f7d59bacc5f38d
which asserts that no search context is created, what do you think?

Reply:
I am good with this, thanks! I will leave it for Jim to review the change itself ;)